File: vl2matplotlib/prepare.py
"""Turn Vega-Lite's *implicit* per-channel `aggregate`/`bin`/`timeUnit` into
explicit pandas statements, mirroring `vl2d3`'s own `prepare.js` role (there
because D3 has no aggregation of its own) -- except here it's less about
hand-rolling the aggregation and more about picking the right one or two
pandas calls, since `groupby(...).agg(...)` already does in one call what
`vl2d3` needs a hand-written `d3.rollup()` reduction for.

Returns a rewritten `encoding` whose channels reference plain, already-
materialized output column names -- `marks.py`/`scales.py` never need to
know aggregate/bin/timeUnit exist at all; they just see flat column names on
`data_var`.
"""

from __future__ import annotations

from .aggops import agg_expr, is_supported_agg_op
from .literals import sanitize_identifier
from .scales import effective_type
from .timeunit import is_supported_timeunit, timeunit_expr


def _rewritten_type(d: dict) -> str | None:
    """The channel's effective type *before* an aggregate/bin/timeUnit
    rewrite below clears the very key (`aggregate`/`bin`/`timeUnit`)
    `effective_type()` would otherwise infer it from -- baked in here as an
    explicit `type` on the rewritten encoding so a downstream consumer
    (chiefly `marks.py`'s bar/tick mark *orientation* inference, which reads
    `is_quantitative()` on the already-rewritten encoding) still sees it.
    Without this, a 1D aggregate bar chart (`x: {aggregate: "sum", field:
    ...}`, no `y`, no explicit `type` anywhere) loses its only quantitative
    signal the moment this module aggregates it, and silently renders as an
    invisible zero-height vertical bar instead of the horizontal one it
    should be."""
    return d.get("type") or effective_type(d)

# Every channel whose own *position* in the mark drawing matters -- x/x2/y/y2
# range companions included, since a binned x2 needs the same bin-derived
# treatment its own x does.
_ALL_CHANNELS = [
    "x", "y", "x2", "y2", "color", "size", "opacity", "shape", "detail", "order", "text",
    "theta", "radius", "xOffset", "yOffset",
]


def _is_pre_binned(bin_val: object) -> bool:
    """Vega-Lite's own "this field is already pre-binned" convention has
    two spellings: the bare string `"binned"`, or the object form
    `{"binned": true, ...}` (which can carry a `step`/other bin-config
    alongside the flag -- `bar_binned_data.vl.json`'s own
    `{"binned": true, "step": 2}` is exactly this shape)."""
    if bin_val == "binned":
        return True
    return isinstance(bin_val, dict) and bin_val.get("binned") is True


def _timeunit_out_name(field: str, unit) -> str:
    name = unit["unit"] if isinstance(unit, dict) else unit
    return sanitize_identifier(f"{name}_{field}")


def _bin_out_names(field: str) -> tuple[str, str]:
    base = sanitize_identifier(field)
    return f"{base}_bin_start", f"{base}_bin_end"


def prepare_encoding(encoding: dict, data_var: str, ignore_unsupported: bool = False) -> tuple[list[str], dict]:
    """Returns `(statements, rewritten_encoding)`."""
    agg_channels = {
        ch: d for ch, d in encoding.items()
        if ch in _ALL_CHANNELS and isinstance(d, dict) and d.get("aggregate") is not None
    }
    bin_channels = {
        ch: d for ch, d in encoding.items()
        # `bin: "binned"` *or* `bin: {"binned": true, ...}` (as opposed to
        # `true`/`{maxbins: ...}`) is Vega-Lite's own "this field is
        # already pre-binned real data -- don't re-bin it" convention: the
        # field IS the bin start, and its own `x2`/`y2` companion (already
        # present in the spec, untouched) is the bin end. Excluded here so
        # it falls through unprepared (no `np.histogram_bin_edges`/`pd.cut`
        # at all) rather than re-binning already-binned data into a
        # different, wrong set of edges -- `bar_binned_data.vl.json`'s own
        # `{"binned": true, "step": 2}` is the dict form of this, easy to
        # miss if only the bare string `"binned"` shorthand is checked for.
        if ch in _ALL_CHANNELS and isinstance(d, dict) and d.get("bin") and not _is_pre_binned(d.get("bin")) and not d.get("aggregate")
    }

    if not agg_channels and not bin_channels:
        return _derive_timeunits_only(encoding, data_var, ignore_unsupported)

    if bin_channels:
        return _prepare_binned(encoding, bin_channels, agg_channels, data_var, ignore_unsupported)

    return _prepare_aggregated(encoding, agg_channels, data_var, ignore_unsupported)


def _derive_timeunits_only(encoding: dict, data_var: str, ignore_unsupported: bool) -> tuple[list[str], dict]:
    stmts: list[str] = []
    rewritten = dict(encoding)
    for ch in _ALL_CHANNELS:
        d = encoding.get(ch)
        if not isinstance(d, dict) or not d.get("field") or not d.get("timeUnit"):
            continue
        unit = d["timeUnit"]
        if not is_supported_timeunit(unit):
            if not ignore_unsupported:
                name = unit["unit"] if isinstance(unit, dict) else unit
                raise ValueError(f"Unsupported timeUnit: {name!r}")
            continue
        out = _timeunit_out_name(d["field"], unit)
        expr = timeunit_expr(unit, f"row[{d['field']!r}]")
        stmts.append(f"{data_var}[{out!r}] = {data_var}.apply(lambda row: {expr}, axis=1)")
        rewritten[ch] = {**d, "field": out}
    return stmts, rewritten


def _prepare_aggregated(encoding: dict, agg_channels: dict, data_var: str, ignore_unsupported: bool) -> tuple[list[str], dict]:
    stmts: list[str] = []
    groupby_fields: list[str] = []
    groupby_channels: list[str] = []
    rewritten = dict(encoding)

    for ch in _ALL_CHANNELS:
        if ch in agg_channels:
            continue
        d = encoding.get(ch)
        if not isinstance(d, dict) or not d.get("field"):
            continue
        field = d["field"]
        if d.get("timeUnit"):
            unit = d["timeUnit"]
            if not is_supported_timeunit(unit):
                if not ignore_unsupported:
                    name = unit["unit"] if isinstance(unit, dict) else unit
                    raise ValueError(f"Unsupported timeUnit: {name!r}")
                continue
            out = _timeunit_out_name(field, unit)
            expr = timeunit_expr(unit, f"row[{field!r}]")
            stmts.append(f"{data_var}[{out!r}] = {data_var}.apply(lambda row: {expr}, axis=1)")
            field = out
            # `_was_timeunit` is an internal-only marker (never a real
            # Vega-Lite encoding key) surviving this rewrite specifically
            # so a downstream consumer can still tell "this field was
            # originally a `timeUnit`, even a *cyclic* single-component one
            # `_rewritten_type()` above had to bake in as plain
            # `"quantitative"` for orientation-inference purposes" -- used
            # by `marks.py`'s own `_render_line_or_area()` to tell a
            # genuine value channel (an `aggregate`) apart from a
            # domain/sequence one that merely happens to read as
            # quantitative too (`area_vertical.vl.json`'s own bare
            # `timeUnit: "year"`, which reduces to a plain int -- see
            # `timeunit.py`'s own module docstring).
            rewritten[ch] = {**d, "field": field, "timeUnit": None, "type": _rewritten_type(d), "_was_timeunit": True}
        if field not in groupby_fields:
            groupby_fields.append(field)
        groupby_channels.append(ch)

    agg_pairs = []  # (out_name, source_field, op)
    for ch, d in agg_channels.items():
        op = d["aggregate"]
        if not is_supported_agg_op(op):
            if ignore_unsupported:
                continue
            raise ValueError(f"Unsupported aggregate op: {op!r}")
        if op == "count":
            out_name = "count"
            source = groupby_fields[0] if groupby_fields else None
        else:
            field = d.get("field")
            if not field:
                if ignore_unsupported:
                    continue
                raise ValueError(f"Unsupported: aggregate {op!r} channel {ch!r} has no field")
            out_name = sanitize_identifier(f"{op}_{field}")
            source = field
        agg_pairs.append((out_name, source, op, ch))

    # Two different encoding channels (e.g. `x` and `order`) aggregating the
    # identical field with the identical op deterministically produce the
    # identical `out_name` -- de-duplicated here (by out_name, first wins)
    # before emitting named-aggregation keyword args / dict keys, since a
    # repeated keyword/key would otherwise be a `SyntaxError`/silently drop
    # a duplicate dict entry. `rewritten` below still needs every channel's
    # own entry, duplicates included, so `agg_pairs` itself stays as-is.
    seen_out_names: set[str] = set()
    unique_agg_pairs = []
    for pair in agg_pairs:
        if pair[0] in seen_out_names:
            continue
        seen_out_names.add(pair[0])
        unique_agg_pairs.append(pair)

    if groupby_fields:
        def _agg_pair(out, src, op):
            if src is None:
                return f"{out}=('__count__', 'size')"
            func = "'size'" if op == "count" else agg_expr(op)
            return f"{out}=({src!r}, {func})"

        named_aggs = ", ".join(_agg_pair(out, src, op) for out, src, op, _ch in unique_agg_pairs)
        if any(src is None for _, src, _, _ in unique_agg_pairs):
            stmts.append(f"{data_var}['__count__'] = 1")
        group_key = groupby_fields[0] if len(groupby_fields) == 1 else groupby_fields
        stmts.append(f"{data_var} = {data_var}.groupby({group_key!r}, as_index=False).agg({named_aggs})")
    else:
        # A groupless aggregate (e.g. a `rule` mark's dataset-wide mean) --
        # `.agg()` on the whole frame returns a scalar per column, not a row,
        # so the result is wrapped into a genuine one-row DataFrame instead.
        parts = []
        for out, src, op, _ch in unique_agg_pairs:
            if op == "count":
                parts.append(f"{out!r}: len({data_var})")
            else:
                parts.append(f"{out!r}: {data_var}[{src!r}].agg({agg_expr(op)})")
        stmts.append(f"{data_var} = pd.DataFrame([{{{', '.join(parts)}}}])")

    for out, _src, _op, ch in agg_pairs:
        d = encoding[ch]
        rewritten[ch] = {**d, "field": out, "aggregate": None, "type": _rewritten_type(d)}

    return stmts, rewritten


def _prepare_binned(encoding: dict, bin_channels: dict, agg_channels: dict, data_var: str, ignore_unsupported: bool) -> tuple[list[str], dict]:
    stmts: list[str] = []
    rewritten = dict(encoding)
    group_keys: list[str] = []

    for bin_ch, bin_def in bin_channels.items():
        field = bin_def["field"]
        max_bins = bin_def["bin"].get("maxbins", 10) if isinstance(bin_def.get("bin"), dict) else 10
        start_col, end_col = _bin_out_names(field)
        # One `__edges` variable *per* bin channel (not a shared name) --
        # 2D binning (e.g. `circle_binned.vl.json`'s own x *and* y both
        # binned) needs two independent edge sets computed from two
        # different fields; a single shared name would have the second
        # channel's own `np.histogram_bin_edges()` call silently overwrite
        # the first's before it's done being used.
        edges_var = f"__edges_{sanitize_identifier(field)}"
        stmts.append(f"{edges_var} = np.histogram_bin_edges({data_var}[{field!r}].dropna(), bins={max_bins})")
        stmts.append(
            f"{data_var}[{start_col!r}] = pd.cut({data_var}[{field!r}], bins={edges_var}, include_lowest=True).apply(lambda iv: iv.left if pd.notna(iv) else float('nan')).astype(float)"
        )
        stmts.append(
            f"{data_var}[{end_col!r}] = pd.cut({data_var}[{field!r}], bins={edges_var}, include_lowest=True).apply(lambda iv: iv.right if pd.notna(iv) else float('nan')).astype(float)"
        )
        rewritten[bin_ch] = {**bin_def, "field": start_col, "bin": None, "type": "quantitative"}
        rewritten[f"{bin_ch}2"] = {"field": end_col, "type": "quantitative"}
        group_keys += [start_col, end_col]

    if agg_channels:
        # Any OTHER channel with its own field but no aggregate of its own
        # (e.g. `color: {field: "Major Genre"}` alongside a binned x and an
        # aggregated y) needs to be a groupby key too, exactly like
        # `_prepare_aggregated`'s own `groupby_fields` collection -- omitting
        # it here would silently drop that column from the aggregated
        # result, breaking any later `.groupby(that_field)` (e.g. the
        # per-color-group draw loop in `marks.py`) with a `KeyError`.
        extra_groupby_fields = []
        for ch in _ALL_CHANNELS:
            if ch in bin_channels or ch in agg_channels:
                continue
            d = encoding.get(ch)
            if isinstance(d, dict) and d.get("field") and d["field"] not in extra_groupby_fields:
                extra_groupby_fields.append(d["field"])
        group_keys += extra_groupby_fields

        agg_pairs = []
        for ch, d in agg_channels.items():
            op = d["aggregate"]
            if not is_supported_agg_op(op):
                if ignore_unsupported:
                    continue
                raise ValueError(f"Unsupported aggregate op: {op!r}")
            if op == "count":
                agg_pairs.append((ch, "count", None, "count"))
            else:
                agg_pairs.append((ch, sanitize_identifier(f"{op}_{d['field']}"), d["field"], op))
        stmts.append(f"{data_var}['__count__'] = 1")

        def _binned_agg_pair(out, src, op):
            func = "'size'" if op is None else agg_expr(op)
            return f"{out}=({(src or '__count__')!r}, {func})"

        named_aggs = ", ".join(_binned_agg_pair(out, src, op) for _ch, out, src, op in agg_pairs)
        stmts.append(
            f"{data_var} = {data_var}.groupby({group_keys!r}, as_index=False, dropna=True).agg({named_aggs})"
        )
        for ch, out, _src, _op in agg_pairs:
            d = encoding[ch]
            rewritten[ch] = {**d, "field": out, "aggregate": None, "type": _rewritten_type(d)}

    return stmts, rewritten

File: vl2plot/src/prepare.js
// Turn Vega-Lite's *implicit* per-channel `aggregate`/`bin`/`timeUnit` into
// (a) data-array statements (for `timeUnit`, which needs a real derived
// field before Plot ever sees the data) and (b) a Plot transform-wrapper
// choice (`Plot.binX`/`binY`/`groupX`/`groupY`) marks.js applies around the
// mark's own already-built options object -- unlike `vl2d3`'s own
// prepare.js (which has to hand-roll the actual grouping/binning via
// `d3.rollup`/`d3.bin`), Plot's own transform functions already do that
// work; this module's job is just deciding *which* one applies and what its
// own `outputs` object should be. Confirmed empirically: `Plot.groupX`/
// `binX` already group by every *other* field-valued channel present in the
// same options object automatically (not just the transform's own named
// axis), matching Vega-Lite's own "every non-aggregate fielded channel is
// an implicit groupby key" semantics for free.

import {isSupportedTimeUnit, timeUnitExpr, isCyclicTimeUnit} from './timeunit.js';
import {isSupportedAggregateOp, plotReducer} from './aggops.js';
import {effectiveType, unescapeFieldName} from './encoding.js';

const POSITION_CHANNELS = ['x', 'y'];
// Every channel Plot's own `groupX`/`groupY` transform can pick up as an
// implicit grouping key (see the module docstring) -- consulted only to
// detect the *degenerate* case where none of them apply at all (see
// `needsConstantKey` below).
const GROUP_KEY_CHANNELS = ['x', 'y', 'color', 'fill', 'stroke', 'opacity', 'size', 'symbol', 'detail'];
// Every channel a mark's own options object can carry a `field` on --
// consulted when deciding whether a channel participates in an implicit
// groupby (see `hasField()` below).
// `detail` (e.g. `detail: {timeUnit: "year", field: "date"}`, a common
// "one line per year" idiom -- repeat_child_layer.vl.json's own shape)
// was previously missing here entirely: it maps onto Plot's own `z`
// channel (marks.js's `commonChannels()`), which still needs its own
// timeUnit applied exactly the same way any other channel's does, but
// silently kept the raw, untruncated field instead -- confirmed
// empirically to leave `z` grouping by the exact (per-row-unique)
// timestamp rather than by year, effectively grouping nothing at all.
const ALL_CHANNELS = ['x', 'y', 'x2', 'y2', 'color', 'fill', 'stroke', 'opacity', 'size', 'symbol', 'r', 'text', 'title', 'detail'];

function fieldRef(rowVar, field) {
  return /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(field) ? `${rowVar}.${field}` : `${rowVar}[${JSON.stringify(field)}]`;
}

// Collects every field that needs `Date` coercion before this mark's own
// channels (or a `timeUnit` deriving from it) can use it -- an explicit
// `type: "temporal"` channel, or a channel whose own `timeUnit` implies
// temporal (a *combined* unit like `yearmonth`; a single cyclic one like
// `month` reduces to a plain number instead, no Date needed as an
// intermediate -- `timeUnitExpr()` still expects a real Date as *input*
// though, so cyclic units need coercion too, just not temporal *output*).
export function collectTemporalFields(encoding) {
  const fields = [];
  for (const ch of ALL_CHANNELS) {
    const def = encoding[ch];
    if (!def || typeof def !== 'object' || typeof def.field !== 'string') continue;
    // Unescaped here (Vega-Lite's own `\.`-means-a-literal-dot convention,
    // e.g. `"a\\.b"` naming the real column "a.b", not a nested "a"->"b"
    // path) -- `renderTemporalCoercion()` (data.js) uses this list's own
    // strings directly as BOTH the read key and the new Date column's own
    // key, so a still-escaped field here reads a column that doesn't
    // exist (`d["a\\.b"]` when the real key is "a.b"), silently coercing
    // to `Invalid Date` for every row (bar_simple_binned_timeunit_
    // special_chars.vl.json's own shape).
    if (def.type === 'temporal' || def.timeUnit) fields.push(unescapeFieldName(def.field));
  }
  return fields;
}

function hasField(def) {
  return def && typeof def === 'object' && typeof def.field === 'string';
}

// Rewrites every `timeUnit`-bearing channel into a plain `field` pointing
// at a newly derived column, returning `{statements, encoding}` --
// `encoding` has every `timeUnit` key stripped (each channel's `field` now
// names the derived output instead), so nothing downstream (`marks.js`,
// `scales.js`) ever needs to know `timeUnit` was involved at all.
export function applyTimeUnits(encoding, dataVar, ignoreUnsupported = false) {
  const statements = [];
  const rewritten = {...encoding};
  const assigns = [];
  for (const ch of ALL_CHANNELS) {
    const def = encoding[ch];
    if (!hasField(def) || !def.timeUnit) continue;
    const unit = def.timeUnit;
    if (!isSupportedTimeUnit(unit)) {
      if (ignoreUnsupported) {
        // Left as the raw (uncoerced) field -- better than dropping the
        // channel outright.
        continue;
      }
      throw new Error(`Unsupported timeUnit: "${JSON.stringify(unit)}"`);
    }
    const unitName = typeof unit === 'object' ? unit.unit : unit;
    // Unescaped for the same reason collectTemporalFields() is, just
    // above -- collectTemporalFields()'s own output (what
    // renderTemporalCoercion() actually creates the coerced Date column
    // under) is ALREADY the real unescaped name, so reading `def.field`
    // here as-is (still escaped) would look up a column that was never
    // created under that exact (backslash-containing) key at all.
    const sourceField = unescapeFieldName(def.field);
    const outField = `${unitName}_${sourceField}`;
    const expr = timeUnitExpr(unit, fieldRef('d', sourceField), ignoreUnsupported);
    assigns.push(`${JSON.stringify(outField)}: ${expr}`);
    const impliedType = isCyclicTimeUnit(unit) ? (def.type && def.type !== 'temporal' ? def.type : 'ordinal') : 'temporal';
    rewritten[ch] = {...def, field: outField, timeUnit: undefined, type: def.type || impliedType};
  }
  if (assigns.length) {
    statements.push(`${dataVar} = ${dataVar}.map(d => ({...d, ${assigns.join(', ')}}));`);
  }
  return {statements, encoding: rewritten};
}

// Decides whether this mark's own (already timeUnit-rewritten) encoding
// needs a `Plot.binX`/`binY`/`groupX`/`groupY` wrapper -- returns `null`
// when no channel carries an inline `bin`/`aggregate` at all (the common
// case), or `{fn, outputs}` otherwise (`fn` is Plot's own transform
// function name; `outputs` is the object its own first argument expects,
// e.g. `{y: "count"}`). Only one axis is ever binned/grouped in Vega-Lite's
// own model (the *other* position channel, plus every additional
// field-valued channel, become implicit groupby keys) -- `x` is preferred
// when both somehow carry `bin`/`aggregate` (a genuinely rare/malformed
// shape), matching this project's own "pick the more common orientation and
// move on" convention elsewhere.
export function planTransform(encoding, ignoreUnsupported = false) {
  let binChannel = null;
  let aggChannel = null;
  for (const ch of POSITION_CHANNELS) {
    const def = encoding[ch];
    if (!def || typeof def !== 'object') continue;
    // `bin: {"binned": true}` OR the equivalent bare-string shorthand
    // `bin: "binned"` (Vega-Lite's own schema allows either -- confirmed
    // against layer_cumulative_histogram.vl.json's own `x: {bin:
    // "binned", ...}`) is Vega-Lite's signal that the field *already
    // holds* the bin boundary (typically alongside an explicit `x2`
    // companion for the other edge) -- the opposite of a request to bin
    // it now, so this deliberately does NOT count as `binChannel` here
    // (a real corpus spec paired this with `x2`, hitting `marks.js`'s own
    // "aggregated value on a continuous bin-interval axis" guard for an
    // aggregate that was never actually being requested). Missing the
    // bare-string form entirely previously misread it as a genuine
    // bin-now request, tripping that exact guard and silently dropping
    // the mark outright under `--ignore-unsupported` (both of this
    // spec's own layers ended up with an empty `marks: []`).
    const isPreBinned = def.bin === 'binned' || (def.bin && typeof def.bin === 'object' && def.bin.binned === true);
    if (def.bin && !isPreBinned) binChannel = binChannel || ch;
    else if (def.aggregate != null) aggChannel = aggChannel || ch;
  }

  if (binChannel) {
    const other = binChannel === 'x' ? 'y' : 'x';
    const otherDef = encoding[other];
    const outputs = {};
    if (otherDef && otherDef.aggregate != null) {
      outputs[other] = plotReducer(otherDef.aggregate, ignoreUnsupported);
    } else {
      outputs[other] = 'count';
    }
    // `Plot.binX`/`binY` accept bin-shaping options (`thresholds`, ...)
    // merged into the same first (`outputs`) argument as the reducers --
    // `bin: {"maxbins": N}`'s closest equivalent is an approximate target
    // bin *count*, which is exactly what Plot's own `thresholds` accepts
    // as a bare number (a precise bin *width*, `bin: {"step": N}`, has no
    // matching Plot option and is left to Plot's own default heuristic).
    // Vega-Lite's own default (`bin: true`, or `bin: {}` with no explicit
    // `maxbins`/`step`) is a fixed `maxbins: 10` -- confirmed against the
    // real compiler's own output for repeat_layer.vl.json (`bin_maxbins_
    // 10_...`, an un-overridden `bin: true`). Left unset, Plot's own
    // auto-threshold heuristic (Sturges/Scott-ish, driven by the data's
    // own size/spread) previously produced a MUCH finer bin count on a
    // large dataset (39 bins on this same spec's ~3000-row movies
    // dataset, not 10) -- a noticeably more jagged, differently-shaped
    // line than every other tool's own 10-bin rendering of the identical
    // spec (the user-reported "doesn't look like the same line plot as
    // the others" symptom).
    const binDef = encoding[binChannel].bin;
    if (binDef && typeof binDef === 'object' && typeof binDef.maxbins === 'number') {
      outputs.thresholds = binDef.maxbins;
    } else if (!(binDef && typeof binDef === 'object' && typeof binDef.step === 'number')) {
      outputs.thresholds = 10;
    }
    return augmentWithTextAggregate({fn: binChannel === 'x' ? 'binX' : 'binY', outputs}, encoding, ignoreUnsupported);
  }

  if (aggChannel) {
    const def = encoding[aggChannel];
    const op = def.aggregate;
    if (op !== 'count' && !isSupportedAggregateOp(op)) {
      if (!ignoreUnsupported) throw new Error(`Unsupported aggregate op: "${op}"`);
    }
    const outputs = {[aggChannel]: plotReducer(op, ignoreUnsupported)};
    // Unlike `binX`/`binY` (named after the axis *being binned*), `groupX`/
    // `groupY` are named after the *grouping key* axis -- the OTHER
    // channel from the one carrying the aggregate (confirmed empirically:
    // `groupX` groups rows by their x value and computes the aggregate
    // output, typically onto y; picking the transform by the aggregate's
    // own channel name, as this used to, silently produced the wrong
    // grouping -- e.g. every row its own bar -- without ever throwing).
    const groupKeyCh = aggChannel === 'x' ? 'y' : 'x';
    // Vega-Lite's genuinely 1-dimensional aggregate -- only the one
    // channel being aggregated is given at all, no other position/color/
    // detail channel to group by (e.g. a single summary bar: `{"x":
    // {"aggregate": "sum", "field": "people"}}`, no `y` at all). Without
    // *some* other channel present, Plot's own group transform has no
    // grouping key to work with at all and silently falls back to each
    // row's own array index -- one "group" per row, not one combined
    // total (confirmed empirically: produced N separate, wrongly-scaled
    // bars instead of a single summed one, without ever throwing). A
    // constant on the missing axis gives every row the same key, which
    // collapses them into the single group Vega-Lite's own semantics call
    // for here.
    //
    // A literal `value`-bound channel (e.g. layer_histogram_global_mean
    // .vl.json's own rule layer: `color: {value: "red"}, size: {value:
    // 5}`, no `y` at all) does NOT count as a usable group key here,
    // despite every row trivially sharing that identical constant --
    // confirmed empirically that Plot's own group transform, given
    // nothing else at all to group by, still falls back to one group
    // PER ROW (the same degenerate case the comment above already
    // describes), silently drawing NO visible rule/mark whatsoever
    // rather than throwing. Only a real per-row FIELD value (which Plot
    // can actually key groups by) counts.
    const hasPositionGroupKey = hasField(encoding[groupKeyCh]);
    // `Plot.groupX`/`groupY` ALSO always group on {z, fill, stroke} first
    // (confirmed from Plot's own source, `transforms/group.js`: both are
    // thin wrappers around a shared `groupn(x, y, ...)` that groups on
    // z/fill/stroke before ever looking at x/y), but they each REQUIRE
    // their own named position option to be present at all -- `options.x
    // == null` throws `"missing channel: x"` at call time (same for y).
    // A rule layer with a real per-row color field but genuinely NO other
    // position channel at all (e.g. layer_line_color_rule.vl.json's own
    // rule layer: `y: {field: "price", aggregate: "mean"}, color: {field:
    // "symbol"}`, no `x` whatsoever) would crash `Plot.groupX` outright --
    // confirmed empirically. `Plot.groupZ` groups purely on {z, fill,
    // stroke} with no x/y requirement, the correct fit here.
    const hasNonPositionGroupKey = GROUP_KEY_CHANNELS.some(ch => ch !== 'x' && ch !== 'y' && ch !== aggChannel && hasField(encoding[ch]));
    let fn, needsConstantKey;
    if (hasPositionGroupKey) {
      fn = aggChannel === 'x' ? 'groupY' : 'groupX';
      needsConstantKey = null;
    } else if (hasNonPositionGroupKey) {
      fn = 'groupZ';
      needsConstantKey = null;
    } else {
      fn = aggChannel === 'x' ? 'groupY' : 'groupX';
      needsConstantKey = groupKeyCh;
    }
    return augmentWithTextAggregate({fn, outputs, needsConstantKey}, encoding, ignoreUnsupported);
  }

  return null;
}

// A `text` channel with its own independent `aggregate` (e.g. a bar
// chart's own label repeating the same summary value alongside its own
// value channel: `{"x": {"aggregate": "sum", "field": "people"}, "text":
// {"aggregate": "sum", "field": "people"}}`) needs its own reducer entry
// in the SAME group/bin transform's `outputs` object, riding along on
// whichever groupby key the position channel(s) already established.
// Left out of `outputs`, Plot's own group/bin transforms would otherwise
// treat the still-raw `text` field (see `marks.js`'s own pairs, which
// keep it as a plain field reference so there's something for this
// reducer to actually read) as an *additional* implicit groupby key --
// every other field-valued channel becomes one, per the module docstring
// -- silently breaking the whole aggregation the same way an unhandled
// `opacity`/`size` channel does (see `marks.js`'s own
// `UNGROUPABLE_STYLE_CHANNELS`); confirmed empirically on a real corpus
// spec (`stacked_bar_h_normalized_labeled`).
function augmentWithTextAggregate(plan, encoding, ignoreUnsupported) {
  const textDef = encoding.text;
  if (!textDef || typeof textDef !== 'object' || textDef.aggregate == null) return plan;
  const op = textDef.aggregate;
  if (op !== 'count' && !isSupportedAggregateOp(op) && !ignoreUnsupported) {
    throw new Error(`Unsupported aggregate op: "${op}"`);
  }
  return {...plan, outputs: {...plan.outputs, text: plotReducer(op, ignoreUnsupported)}};
}
